quixstreams/state/base/transaction.py

```diff
@@ -25,24 +15,131 @@
+    to simplify the querying over them.
+    """
+
+    def __init__(self):
+        # A map with updated keys in format {<cf>: {<prefix>: {<key>: <value>}}}
+        # Note: "updates" are bucketed per prefix to speed up iterating over the
+        # specific set of keys when we merge updates with data from the stores.
+        # Using a prefix like that allows us to perform fewer iterations.
+        self._updated: dict[str, dict[bytes, dict[bytes, bytes]]] = defaultdict(
+            lambda: defaultdict(dict)
+        )
+        # Dict of sets with deleted keys in format {<cf>: set[<key1>, <key2>]}
+        # Deletes are stored without prefixes because we don't need to iterate over
+        # them.
+        self._deleted: dict[str, set[bytes]] = defaultdict(set)
+        self._empty = True
+
+    def get(
+        self,
+        key: bytes,
+        prefix: bytes,
+        cf_name: str = "default",
+    ) -> Union[bytes, Undefined]:
+        """
+        Get a value for the key.
+
+        Returns the key value if it has been updated during the transaction.
+
+        If the key has already been deleted, returns "DELETED" sentinel
+        (we don't need to check the actual store).
+        If the key is not present in the cache, returns "UNDEFINED sentinel
+        (we need to check the store).
+
+        :param: key: key as bytes
+        :param: prefix: key prefix as bytes
+        :param: cf_name: column family name
+        """
+        # Check if the key has been deleted
+        deleted = key in self._deleted[cf_name]
+        if deleted:
+            # The key is deleted and the store doesn't need to be checked
+            return DELETED
+
+        # Check if the key has been updated
+        # If the key is not present in the cache, we need to check the store and return
+        # UNDEFINED to signify that
+        cached = self._updated[cf_name][prefix].get(key, UNDEFINED)
+        return cached
+
+    def set(self, key: bytes, value: bytes, prefix: bytes, cf_name: str = "default"):
+        """
+        Set a value for the key.
+
+        :param: key: key as bytes
+        :param: value: value as bytes
+        :param: prefix: key prefix as bytes
+        :param: cf_name: column family name
+        """
+        self._updated[cf_name][prefix][key] = value
+        self._deleted[cf_name].discard(key)
+        self._empty = False
+
+    def delete(self, key: Any, prefix: bytes, cf_name: str = "default"):
+        """
+        Delete a key.
+
+        :param: key: key as bytes
+        :param: value: value as bytes
+        :param: prefix: key prefix as bytes
+        :param: cf_name: column family name
+        """
+        self._updated[cf_name][prefix].pop(key, None)
+        self._deleted[cf_name].add(key)
+        self._empty = False
+
+    def is_empty(self) -> bool:
+        """
+        Return True if any changes have been made (updates or deletes), otherwise
+        return False.
+        """
+        return self._empty
```

Comment:
I wanted to do it this way at first, but to say whether `_updated` is empty or not, you need to iterate over it.
Due to `defaultdict`, `.get()` will create a key on the first prefix level, but the actual key may not be set.

Using a `self._empty` was easier because if there's at least one "delete" or "update", the cache is not empty anymore (even if the updated key was deleted).